### ISSUE TITLE
Passing complete Objects

### ISSUE THREAD

**Issue by MarvinTeichmann** [TensorVision/TensorVision#54]
Currently TV generically builds Tensorflow Models of the following shape:

```
# Get Inputs
images, labels = data_input.inputs(hypes, q[phase])

# Run inference on the encoder network
logits = encoder.inference(hypes, images)

# Build decoder on top of the logits
decoded_logits = objective.decoder(hypes, logits)

# Compute losses
losses = objective.loss(hypes, decoded_logits, labels)


# Build training operation
train_op = optimizer.training(hypes, losses,
                              global_step, learning_rate)

# Write Values to summary
eval_op = objective.evaluation(hypes, images, labels,
                               decoded_logits, losses)
```

This is great, as (almost) all deep computer vision models are designable in this way. I have however observed a flaw with it: Each step in the building process (i.e. `objective.decoder(hypes, logits)` ) only has access to the last tensor of the computational graph (`logits`). 

I have seen situations where this is not enough. In particular:
1) To build a decoder which uses information of more layers (e.g.  `logits`, `pool5` and `pool3`)
2) Allow evaluation to do see the current `step`. (i.e. to do more accurate evaluation every 1000 iteration).

Most of it can be solved by `abusing` the current interface. For example, to give `decoder` access to more layers, one could:
1) Call `tf.get_default_graph()` and read all operations from there <sup>1</sup>.
2) Make `logits` a list/tuple of tensors. (e.g. `logits = [logits, pool5, pool3]` ).
3) Abose the `hypes` object, to pass tensors (e.g. call `hypes['network']['pool5'] = pool5) to make pool5 easily accessible in all later layers. (Hypes is basically passed to all feature functions).

<sup>1</sup> [Note, that a `tensor object` like `pool5` is infect only a `link` to an operation in the tensorflow graph. The tensor itself is therefore not lost, if it gets overwritten or not passed on and can be recovered.]

This idea has worked in all  _critical_ situations and I think one could say, that any graph can be build in this structure, as the overall framework allows to pass any information to further functions. However I am not sure, whether we should make this sort of abuse intentionally. Some of the solutions, like reading from the tf.graph directly are pretty hacky and should imho be avoided.

Additionally some less critical variables of tensorVision, like the `learning rate` or the `current steps` and the `summary writer` are never passed to any user defined functions. Allowing access to those parameters would require to adopt the interface. So if we want to allow `objective.evaluation` to see the current `step` we would need to pass this directly. Which then has the disadvantage, that any feature  `objective.evaluation` implementation needs to be able to accept the parameter `step`. 

---

As a conclusion, I would like to allow each TensorVision function to see all important parameters. I see three main groups of those parameters:
1) hypes (already accessible by all functions) 
2) tv-graph (i.e. object of already defined Tensors)
3) tv-session (basically the output of `tv.core.start_tv_session(hypes`); containing tensorflow objects like `saver` `summary writer` and `sess`. )

All three groups could aether be implemented as `dict` or as python `object`. Where the python object implementation would look similar to [our vgg-fcn](https://github.com/MarvinTeichmann/tensorflow-fcn) implementation: `encoder.inference` could create the object and all further functions just append tensors to it. What do you think is preferred?

Additionally, I see three ways of passing those objects to the modules:
1) Pass them explicitly. That is `objective.loss` would become `objective.loss(hypes, tv-graph, tv-session, decoded_logits, labels)`.
2) Boundle the information in one object. For example we can build an object/dict `tv` with members `tv.hypes`, `tv.graph` and `tv.session`.
3) Build a getter function, which allows to access `tv-graph` from anywhere (i.e. similar to `tf.get_default_graph()`).

**Comment by MarvinTeichmann** [comment]
Sorry for posting such a very long issue. 
@MartinThoma To you have any thoughts about all that?

**Comment by MartinThoma** [comment]
I think (2) (Make logits a list/tuple of tensors. ) doesn't sound too bad.

> Should we use an object or a dictionary?

Objects (or classes) are used when you need some logic. When you only need to store data (which can be objects), then a dictionary should be used. 

> Additionally, I see three ways of passing those objects to the modules

As you want to pass it to all functions, you could also use a global object. Or better a module. See: https://docs.python.org/2/faq/programming.html#how-do-i-share-global-variables-across-modules
